src/pudl/transform/ferc1.py

```diff
@@ -960,75 +1124,332 @@ def calculate_values_from_components(
-                f"Calculations in {table_name} are off by {off_ratio:.2%}. Expected tolerance "
-                f"of {calculation_tolerance:.1%}."
+                f"The table is missing the following required columns: {missing_required_cols}"
             )
 
-    # We'll only get here if the proportion of calculations that are off is acceptable
-    if (off_ratio > 0 or np.isnan(off_ratio)) and add_corrections:
-        logger.info(
-            f"{table_name}: has {len(off_df)} ({off_ratio:.02%}) records whose "
-            "calculations don't match. Adding correction records to make calculations "
-            "match reported values."
-        )
-        corrections = off_df.copy()
-        corrections[value_col] = (
-            corrections[value_col].fillna(0.0) - corrections["calculated_amount"]
-        )
-        corrections["original_factoid"] = corrections["xbrl_factoid"]
-        corrections["xbrl_factoid"] = corrections["xbrl_factoid"] + "_correction"
-        corrections["row_type_xbrl"] = "correction"
-        corrections["is_within_table_calc"] = False
-        corrections["record_id"] = pd.NA
-
-        calculated_df = pd.concat(
-            [calculated_df, corrections], axis="index"
-        ).reset_index()
-    return calculated_df
+    @abstractmethod
+    def metric(self: Self, gb: DataFrameGroupBy) -> pd.Series:
+        """Apply method for each individual test."""
+        ...
+
+    def is_not_close(self, df: pd.DataFrame) -> pd.Series:
+        """Flag those records that qualify as an error.
+
+        NOTE: How should this categorization vary from metric to metric? Why is it
+        reasonable to only look at records where ``abs_diff`` is not Null?
+        """
+        return pd.Series(
+            ~np.isclose(
+                df["calculated_value"],
+                df["reported_value"],
+                rtol=self.metric_inputs.isclose_rtol,
+                atol=self.metric_inputs.isclose_atol,
+            )
+            & df["abs_diff"].notnull()
+        )
+
+    def groupby_by(self: Self) -> list[str]:
+        """The list of columns to group by.
+
+        We want to default to adding the table_name into all groupby's, but two of our
+        ``by`` options need special treatment.
+        """
+        gb_by = ["table_name", self.by]
+        if self.by in ["ungrouped", "table_name"]:
+            gb_by = [self.by]
+        return gb_by
+
+    def calculate_metric(self: Self, df: pd.DataFrame) -> pd.Series:
+        """Calculate the frequency with which records are tagged as errors."""
+        df["is_not_close"] = self.is_not_close(df)
+        return df.groupby(by=self.groupby_by()).apply(self.metric)
+
+    def snake_case_metric_name(self: Self) -> str:
+        """Convert the TitleCase class name to a snake_case string."""
+        class_name = self.__class__.__name__
+        return re.sub("(?!^)([A-Z]+)", r"_\1", class_name).lower()
+
+    def check(self: Self, calculated_df) -> pd.DataFrame:
```

Comment:
I think `check()` is probably a good name for this method, if the classes are named after the error metric then we'll have things like:

```py
# the generic case
error_metric.check(calculated_df)
error_frequency.check(calculated_df)
relative_error_magnitude.check(calculated_df)
```

which seems legible.